go/oasis-test-runner/scenario/e2e/governance_upgrade.go

```diff
@@ -0,0 +1,379 @@
+package e2e
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"sync"
+	"time"
+
+	"github.com/oasisprotocol/oasis-core/go/common/crypto/hash"
+	"github.com/oasisprotocol/oasis-core/go/common/quantity"
+	consensus "github.com/oasisprotocol/oasis-core/go/consensus/api"
+	"github.com/oasisprotocol/oasis-core/go/consensus/api/transaction"
+	epoch "github.com/oasisprotocol/oasis-core/go/epochtime/api"
+	"github.com/oasisprotocol/oasis-core/go/governance/api"
+	"github.com/oasisprotocol/oasis-core/go/oasis-test-runner/env"
+	"github.com/oasisprotocol/oasis-core/go/oasis-test-runner/log"
+	"github.com/oasisprotocol/oasis-core/go/oasis-test-runner/oasis"
+	"github.com/oasisprotocol/oasis-core/go/oasis-test-runner/scenario"
+	staking "github.com/oasisprotocol/oasis-core/go/staking/api"
+	upgrade "github.com/oasisprotocol/oasis-core/go/upgrade/api"
+	"github.com/oasisprotocol/oasis-core/go/upgrade/migrations"
+)
+
+var (
+	// GovernanceConsensusUpgrade is the governance consensus upgrade scenario.
+	GovernanceConsensusUpgrade scenario.Scenario = newGovernanceConsensusUpgradeImpl(true)
+	// GovernanceConsensusFailUpgrade is the governance consensus upgrade scenario where
+	// node should fail the upgrade.
+	GovernanceConsensusFailUpgrade scenario.Scenario = newGovernanceConsensusUpgradeImpl(false)
+	// TODO: cancel upgrade test case.
+	// TODO: governance state in genesis.
+)
```

Comment:
Would be cool to also have a test with runtimes surviving a consensus layer upgrade.

Reply:
Done, i moved the tests under `e2e/runtime/` and the tests now also exercise the runtime before and after the upgrade.